diff --git a/lib/internal/module.js b/lib/internal/module.js
index 2f38618d..f9f2d021 100644
--- a/lib/internal/module.js
+++ b/lib/internal/module.js
@@ -17,7 +17,13 @@ function makeRequireFunction() {
   function require(path) {
     try {
       exports.requireDepth += 1;
-      return self.require(path);
+
+      // @F8CHANGE
+      if (global.f8 && global.f8.require) {
+        return global.f8.require(path, self);
+      }
+
+      throw 'F8 require function not found';
     } finally {
       exports.requireDepth -= 1;
     }
