diff --git a/lib/module.js b/lib/module.js
index d485c7c9..d00c6d6d 100644
--- a/lib/module.js
+++ b/lib/module.js
@@ -419,12 +419,12 @@ Module._load = function(request, parent, isMain) {
 
   var cachedModule = Module._cache[filename];
   if (cachedModule) {
-    return cachedModule.exports;
+    return maskNative(parent, filename, cachedModule.exports);
   }
 
   if (NativeModule.nonInternalExists(filename)) {
     debug('load native module %s', request);
-    return NativeModule.require(filename);
+    return maskNative(parent, filename, NativeModule.require(filename));
   }
 
   var module = new Module(filename, parent);
@@ -441,6 +441,23 @@ Module._load = function(request, parent, isMain) {
   return module.exports;
 };
 
+function maskNative(parent, id, exports) {
+  // if a builtin is requiring another builtin, let it through
+  if (NativeModule.nonInternalExists(parent.id) && NativeModule.nonInternalExists(id)) {
+    return exports;
+  }
+
+  // if a userspace mod is requiring a system builtin that has been overridden, then return the override
+  if (global._j2v8 && global._j2v8.maskNative) {
+    let maskModule = global._j2v8.maskNative(id, exports);
+    if (maskModule) {
+      return maskModule;
+    }
+  }
+
+  return exports;
+}
+
 function tryModuleLoad(module, filename) {
   var threw = true;
   try {
